content/posts/notes-on-2pc.md

```diff
@@ -5,38 +5,73 @@ date: 2019-09-04
 
-### 2PC applies to databases only
-2PC tells nothing about resources participating in the protocol. In MS DTC those could be databases, queueing systems, Web Services etc.
+This is not "yet another introduction to 2PC"(TM). If you need a refresher read one of many descriptions out there[^1] before continuing.
 
-### 2PC provides atomic visibility
-2PC is an atomic commit protocol meaning that all participants will eventually commit. It **does not** provide atomic visibility i.e. making sure that it's impossible to see partial commit results.  
+### 2PC does not provide "transactions"
+2PC is an atomic commit protocol meaning all participants will **eventually commit** if all voted "YES" - No more, no less than that. 
 
-Two databases in snapshot isolation mode and reads from one and the other. 
+Let's look at an example to see what we mean by "no transactions". In our scenario we have two participants: a database and a messages queue. 
 
-Indicate difference between MS DTC (with default configuration) and 2PC.
+{{< figure src="/posts/2pc-atomic-visibility-scenario.jpg" title="2PC atomic visibility">}}
 
-### 2PC does not tolerate failures
-Participants can't make progress *on-their-own* when waiting for `Coordinators` decision. That being said that *window* of vulnerability is smaller than some might think (only during voting not during performing transactional work on each resource).
+The diagram shows the part of the 2PC protocol after both participants voted "YES" and the cooridantor is commiting.  In our example, the queue transactoins commits first however 2PC says nothing about order in which the participants commit or the delays between the commits. It's undeterministic in other words.
 
-Coordinator is a single point of failure but we can make it more fault tolerant! e.g. spanner's case. 
+What's most interesting is the ouside observer i.e. the client. It makes a read requests to both participants. In case of the message queue, the read request arrives after the commit from the coordinator. This means that the read operation will return any messages send to the queue as part of the transaction that just comitted. 
 
-Talking of implementation: "Is your DB more fault tolerant than your transaction coordinator?".
+In case of the database the read request arrives before the commit. What will be the result here? First, 2PC says nothing about read behavior - it's outside of the system model defined by the protocol. The only requirement is that the database guarantees successful commit. 
 
-### 2PC's biggest problem is commit latency
-It's not the latency but 
+The result of the read operation depends on the deployment configuration and there are at least two possible behaviors. The the read operation can:
+    
+* Hang until the local transaction is committed - this will happen when local transaction operates in `Serializable` isolation level. This is default with MSDTC and MS SqlServer unless explictly changed,
```

Comment:
```suggestion
* Block until the local transaction is committed - this will happen when local transaction operates in `Serializable` isolation level. This is default with Microsoft Distributed Transaction Coordinator (MSDTC, an implementation of 2PC built into Windows) and Microsoft SQL Server, but can be changed on a per-transaction basis.
```